modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Iri.scala

```diff
@@ -970,4 +1019,36 @@ object Iri {
       case url: Url       => url.show
       case urn: Urn       => urn.show
     }
+
+  private val `sub-delims` = Set('!', '$', '&', '\'', '(', ')', '*', '+', ',', ';', '=').map(_.toByte)
+  private val unreserverd = Set('-', '_', '.', '~').map(_.toByte) ++ ('a' to 'z').map(_.toByte).toSet ++ ('A' to 'Z')
+    .map(_.toByte)
+    .toSet ++ ('0' to '9').map(_.toByte).toSet
+  private val pchar = `sub-delims` ++ unreserverd + ':'.toByte + '@'.toByte
+
+  private def pctEncode(s: String) = {
+    val in = s.getBytes(UTF_8)
+    in.foldLeft(new StringBuilder()) {
+        case (sb, b) if pchar.contains(b) => sb.append(b.toChar)
+        case (sb, b)                      => sb.append("%").append(upperHexDigit((b >> 4) & 0xF)).append(upperHexDigit(b & 0xF))
+      }
+      .toString()
+  }
+
+  private[rdf] trait PctString {
```

Comment:
I think we should take out the PctString into a separate file, and move the constants in the companion object.